reconcile/external_resources/integration_secrets_sync.py

```diff
@@ -0,0 +1,32 @@
+from reconcile.external_resources.model import ExternalResourcesInventory
+from reconcile.external_resources.secrets_sync import VaultSecretsReconciler
+from reconcile.typed_queries.app_interface_vault_settings import (
+    get_app_interface_vault_settings,
+)
+from reconcile.typed_queries.external_resources import get_namespaces, get_settings
+from reconcile.utils.openshift_resource import ResourceInventory
+from reconcile.utils.secret_reader import create_secret_reader
+from reconcile.utils.semver_helper import make_semver
+
+QONTRACT_INTEGRATION = "external_resources_secrets_sync"
+QONTRACT_INTEGRATION_VERSION = make_semver(0, 1, 0)
+
+
+def run(dry_run: bool, thread_pool_size: int) -> None:
+    """Integration that syncs External Resources Outputs Secrets from Vault into
+    the target clusters
+    """
+    vault_settings = get_app_interface_vault_settings()
+    secrets_reader = create_secret_reader(use_vault=vault_settings.vault)
+    er_settings = get_settings()[0]
+    namespaces = [ns for ns in get_namespaces() if ns.external_resources]
+    er_inventory = ExternalResourcesInventory(namespaces)
+
+    reconciler = VaultSecretsReconciler(
+        ri=ResourceInventory(),
+        secrets_reader=secrets_reader,
+        vault_path=er_settings.vault_secrets_path,
+        thread_pool_size=thread_pool_size,
+        dry_run=dry_run,
+    )
+    reconciler.sync_secrets(er_inventory.values())
```

Comment:
Could there be any conflict operations caused by concurrent run of this integration and the external resources integration?

Reply:
It can happen (rarely) that a secret is written by this integration right after the external resources integration writes the secret to the cluster and before writing it to Vault. The last secret will eventually land in the cluster with this integration. 

If this becomes a problem we can think of some optimistic locking approach or other mechanism, but I rather wait to see if this becomes an issue.

Reply:
my understanding of current flow of external resources integration is to sync secrets to cluster first, then save to vault, is it correct?

consider this flow:

* external resources integration sync secret v1 to cluster
* external resources integration save secret v1 to vault
* secrets sync integration sync secret v1 (no op)

Then new changes come in:

* external resources integration sync secret v2 to cluster
* secrets sync integration sync secret v1 to cluster (because vault not updated yet)
* external resources integration save secret v2 to vault
* next secrets sync integration sync secret v2 to cluster

So eventually the secret is correct for this case.

What if some errors happen when the integration write to vault failed after successful sync secrets to cluster?

Reply:
Yup, I see where are you going. 
I think the easier solution for this is to introduce an intermediate state (like PENDING_SECRET_SYNC) before marking a resource as complete. Then if the secret is written successfully, mark it as CREATED. 
This way the main integration would retry the sync logic until it finishes successfully. WDYT?

Reply:
How about save to vault first then sync to cluster? Then cluster secrets should always update after vault value update, reduce the chance of sync old secrets to cluster.

Reply:
But the main problem still persists. 
If a resource is updated (Terraform) then the Secret needs to land into the cluster ASAP (RDS password for example). 

* If the Vault call fails, the secret will eventually be synced with a wrong value. 
* If the cluster OC fails, the secret can take too much time to land in a cluster 

So to improve this I propose to: 
* Use an additional STATE (e:g PENDING_SECRET_SYNC) that will be set in case the secret sync fails (either OC or VAULT) 
* Secrets Sync integration will query the STATE and only sync resources marked as CREATED. 
* The main integration will retry to sync PENDING_SECRET_SYNC resources at each iteration. Since the intermediate Secret will exist there is no need to run terraform on resources with this state.

Reply:
I'm thinking can we do it without introducing a shared state between 2 integrations.

The requirements are:

1. sync secrets to cluster ASAP
2. eventually write secrets to vault
3. old secrets shouldn't overwrite new secrets

The last action of in cluster reconcile is to delete source secret, can we use it to indicate `PENDING_SECRET_SYNC`? Then requirement 1 and 2 are met.

To avoid old secrets overwrite new secrets, we can use optimistic lock to cover most cases, add a new version field (can be timestamp) to secret, before apply secret to cluster, check if desired version is greater than current version

Reply:
I introduced the optimistic locking using a timestamp + the PENDING_SECRET_SYNC state in case the secret sync operation fails for a spec. PTAL.

Reply:
the new state `PENDING_SECRET_SYNC` is treated as an error state added later, `CREATED` is the final success state, the current flow is state saved as `CREATED` after resource applied, mark to `PENDING_SECRET_SYNC` if failed, then finally change back to `CREATED` if sync success, this makes the code a little complex as need to do different state transition in https://github.com/app-sre/qontract-reconcile/pull/4375/files#diff-a7fbe2fedf2aacf47fbbe401c068e86e355ad64676b7ca07367c7dfcc4ce800bR278.

How about change the state flow to after resource applied, if it requires secret sync, save state as `PENDING_SECRET_SYNC`, then change it to `CREATED` after sync?

This way you can just pass all keys require sync to `_sync_secrets` as single param, mark them to `CREATED` if no errors.

Reply:
I wanted to avoid an additional DDB update for each spec reconciliation. Since PENDING_SECRET_SYNC is an error state it should not be happening assiduously. 
But I agree that the flow is better understood this way. I changed it to your suggestion.

Reply:
We can improve update performance later if needed, like use batch update instead of update one by one to reduce requests